src/include/hardware.h

```diff
@@ -25,6 +25,23 @@
         WT_READ_BARRIER();      \
     } while (0)
 
+/*
+ * In some architectures with weak memory ordering, the CPU can reorder the reads across full
+ * barriers in other threads. Guarantee that subsequent reads do not see any earlier state in those
+ * architectures.
+ */
+#define WT_ORDERED_READ_FOR_WEAK_MEMORY_ORDERING_ARCH(v, val) \
+    do {                                                      \
+        (v) = (val);                                          \
+        WT_READ_BARRIER_FOR_WEAK_MEMORY_ORDERING_ARCH();      \
+    } while (0)
+
+/* Ensure the value is only read once using a compiler barrier. */
+#define WT_READ_ONCE(v, val) \
+    do {                     \
+        (v) = (val);         \
+        WT_BARRIER();        \
```

Comment:
Here we don't necessarily need the compiler barrier. As we only intend to ensure the variable is read once, we can do the same thing as linux's READ_ONCE macro to cast the value to volatile. But I don't want to be too aggressive in this pr and I think we can go with a compiler barrier for now.

Reply:
Agreed, let's keep this as simple as possible while still solving the problem.

Reply:
I agree too - I don't want to introduce volatile at the moment.